iree/compiler/Dialect/Flow/IR/FlowOpFolders.cpp

```diff
@@ -567,7 +567,8 @@ static uint64_t getFlattenedIndex(ShapedType type, ArrayRef<uint64_t> index) {
 
 static bool compareShapesEqual(ShapedType lhsType, ValueRange lhsDynamicDims,
                                ShapedType rhsType, ValueRange rhsDynamicDims) {
-  if (lhsType.hasStaticShape() && lhsType == rhsType) {
+  if (lhsType.hasStaticShape() &&
+      lhsType.getNumElements() == rhsType.getNumElements()) {
```

Comment:
If they are static and have the same rank, checking type would be good enough. I believe that they should have same rank because the function name is `compareShapesEqual` and I added the check below.

Reply:
Actually you can fold it even if the rank is not the same since the number of elements is the same. I know the check below is needed for dynamic case check that follows, but I think that might need reworking. Need to revisit later.